File: ch02/productinfo/java/client/src/main/java/ecommerce/ProductInfoClient.java
package ecommerce;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

import java.util.logging.Logger;

/**
 * gRPC client sample for productInfo service.
 */
public class ProductInfoClient {

    private static final Logger logger = Logger.getLogger(ProductInfoClient.class.getName());

    public static void main(String[] args) throws InterruptedException {
        ManagedChannel channel = ManagedChannelBuilder.forAddress("localhost", 50051)
                .usePlaintext()
                .build();

        ProductInfoGrpc.ProductInfoBlockingStub stub =
                ProductInfoGrpc.newBlockingStub(channel);

        ProductInfoOuterClass.ProductID productID = stub.addProduct(
                ProductInfoOuterClass.Product.newBuilder()
                        .setName("Samsung S10")
                        .setDescription("Samsung Galaxy S10 is the latest smart phone, " +
                                "launched in February 2019")
                        .setPrice(700.0f)
                        .build());
        logger.info("Product ID: " + productID.getValue() + " added successfully.");

        ProductInfoOuterClass.Product product = stub.getProduct(productID);
        logger.info("Product: " + product.toString());
        channel.shutdown();
    }
}

File: ch02/productinfo/kotlin/client/src/main/kotlin/ecommerce/ProductInfoClient.kt
package ecommerce

import ecommerce.ProductInfoGrpcKt.ProductInfoCoroutineStub
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asExecutor
import kotlinx.coroutines.runBlocking
import java.io.Closeable
import java.util.concurrent.TimeUnit

class ProductInfoClient (
    private val channel: ManagedChannel,
) : Closeable {
  private val stub: ProductInfoCoroutineStub = ProductInfoCoroutineStub(channel)

  suspend fun addProduct(name: String, description: String, price: Float): String {
    val product = product {
      this.name = name
      this.description = description
      this.price = price
    }

    val response = stub.addProduct(product)
    val productID = response.value
    println("Product ID:  $productID added successfully.")
    return productID
  }

  suspend fun getProduct(productID: String): Product {
    val request = productID {
      value = productID
    }
    val response = stub.getProduct(request)
    println("Product: $response")
    return response
  }

  override fun close() {
    channel.shutdown().awaitTermination(5, TimeUnit.SECONDS)
  }
}


fun main() = runBlocking {
  val port = 50051

  val client = ProductInfoClient(
        ManagedChannelBuilder.forAddress("localhost", port)
            .usePlaintext()
            .executor(Dispatchers.Default.asExecutor())
            .build())


  val productID = client.addProduct("Samsung S10", "Samsung Galaxy S10 is the latest smart phone, " +
          "launched in February 2019", 700.0f)
  val product = client.getProduct(productID)
  client.close()
}